cpp/src/arrow/type.h

```diff
@@ -677,7 +678,8 @@ class ARROW_EXPORT DictionaryType : public FixedWidthType {
 
 class ARROW_EXPORT Schema {
  public:
-  explicit Schema(const std::vector<std::shared_ptr<Field>>& fields);
+  explicit Schema(const std::vector<std::shared_ptr<Field>>& fields,
+      const KeyValueMetadata& custom_metadata = KeyValueMetadata());
```

Comment:
Hm, this seems like a lot of additional complexity relative to what the purpose of this class is. What do we get from using `unique_ptr`?